api/hypershift/v1alpha1/nodepool_types.go

```diff
@@ -888,16 +888,55 @@ type AgentNodePoolPlatform struct {
 	AgentLabelSelector *metav1.LabelSelector `json:"agentLabelSelector,omitempty"`
 }
 
-type AzureNodePoolPlatform struct {
-	VMSize string `json:"vmsize"`
-	// ImageID is the id of the image to boot from. If unset, the default image at the location below will be used:
-	// subscription/$subscriptionID/resourceGroups/$resourceGroupName/providers/Microsoft.Compute/images/rhcos.x86_64.vhd
+type AzureVMImageType string
+
+const (
+	ImageID          AzureVMImageType = "ImageID"
+	AzureMarketplace AzureVMImageType = "AzureMarketplace"
+)
+
+// AzureVMImage represents the different types of image data that can be provided for an Azure VM.
+// +union
+type AzureVMImage struct {
+	// Type is the type of image data that will be provided to the Azure VM. This can be either "ImageID" or
+	// "MarketplaceImageInfo".
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=ImageID;AzureMarketplace
+	// +unionDiscriminator
+	Type AzureVMImageType `json:"azureImageType"`
+
+	// ImageID is the Azure resource ID of a VHD image to use to boot the Azure VMs from.
+	//
 	// +optional
+	// +unionMember, optional
 	ImageID string `json:"imageID,omitempty"`
+
+	// MarketplaceImageInfo contains the Azure Marketplace image info to use to boot the Azure VMs from.
+	//
+	// +optional
+	// +unionMember, optional
+	MarketplaceImageInfo *MarketplaceImage `json:"marketplaceImage,omitempty"`
+}
+
+type AzureNodePoolPlatform struct {
+	// VMSize is the Azure VM instance type to use for the nodes being created in the nodepool.
+	//
+	// +kubebuilder:validation:Required
```

Comment:
fwiw everything is required unless told otherwise. Shouldn't we have a len>0 validation here?

Reply:
> fwiw everything is required unless told otherwise.

The rules for this are a little hairy and depend on package level markers and whether something is `omitempty` or not, so we normally recommend that every API field is marked explicitly as required or optional, we have had APIs accidentally switch required fields to optional in the past because of these non-obvious rules

Reply:
The criteria in this pkg it's been all is required unless mark as optional.
If we want to be explicit which I agree we should, let's follow up to do it consistently for all the API.

Reply:
Created a story for our backlog - [HOSTEDCP-1825](https://issues.redhat.com/browse/HOSTEDCP-1825)